### ISSUE TITLE
Knife Support for SSH Config File

### ISSUE THREAD

**Issue by darkn3rd** [chef/chef#7053]
## Description

Knife doesn't support ssh-config file with SSH.   This is ***enhancement request*** to support it.

Currently, one would have to type something like this:

```bash
knife bootstrap localhost \
  --ssh-user vagrant \
  --sudo 
  --identity-file .vagrant/machines/tools.dev/virtualbox/private_key \
  --ssh-port 2222 
  --node-name tools.dev 
  --ssh-port PORT
```

If ssh-config file was supported, we could do this:

```bash
knife bootstrap --ssh-file vm_guests.ssh-config --node-name tools.dev 
```

## Chef Version

* **host system**: `Chef: 13.4.19`
* **guest system**: `Chef: 13.8.5`

## Platform Version

* **host system**: Ubuntu 16.04
* **guest system**: Ubuntu 14.04

**Comment by darkn3rd** [comment]
@coderanger I am not sure what you mean?

My ssh-config format would look like this:

```
Host tools.dev
  HostName 127.0.0.1
  User vagrant
  Port 2222
  UserKnownHostsFile /dev/null
  StrictHostKeyChecking no
  PasswordAuthentication no
  IdentityFile ~/myproject/.vagrant/machines/tools.dev/virtualbox/private_key
  IdentitiesOnly yes
  LogLevel FATAL

Host es-01.dev
  HostName 127.0.0.1
  User vagrant
  Port 2200
  UserKnownHostsFile /dev/null
  StrictHostKeyChecking no
  PasswordAuthentication no
  IdentityFile ~/myproject/.vagrant/machines/es-01.dev/virtualbox/private_key
  IdentitiesOnly yes
  LogLevel FATAL
```

**Comment by coderanger** [comment]
@darkn3rd You can use the matching feature to streamline things a bit, `Host *.dev`.

**Comment by darkn3rd** [comment]
I got that.  Vagrant auto-generates this.  This ticket was about supporting the ssh-config file format that the underlying ruby library supports, but the knife tool doesn't.

**Comment by darkn3rd** [comment]
The underlying `Net::SSH::Config` load method has support for this:

* https://net-ssh.github.io/net-ssh/classes/Net/SSH/Config.html

So the only changes would be to have a option for it, call that method to load the configuration.

**Comment by jwadolowski** [comment]
It seems I've just encountered 2 use cases that `~/.ssh/config` support could solve right off the bat.

## Signal propagation to remote hosts

Our deployment pipeline that's currently in place works as follows:
* deployment is defined as a Jenkins job
* `sh` step triggers `knife ssh "QUERY" "DEPLOYMENT_CMD" --ssh-user XYZ --ssh-identity-file /path/to/.ssh/id_rsa --exit-on-error`
* `DEPLOYMENT_CMD` is a wrapper script on top of `chef-client`

The problem occurs if deployment pipeline is aborted for some reason. Jenkins sends `SIGTERM` to `knife` and... that's basically it. `knife` gets terminated, SSH connection closes immediately and `DEPLOYMENT_CMD` never knows that such action has been taken.

To propagate signals pseudo terminal has to be allocated when establishing SSH connection. When `ssh` command is used then everything works fine:

```
$ ssh -tt user@host "DEPLOYMENT_CMD"
[INFO] Starting chef-client wrapper
[INFO] <other things>
Starting Chef Client, version 14.11.21
resolving cookbooks for run list: [ ... ]
Synchronizing Cookbooks:
  - apt (7.0.0)
  - ulimit (1.0.0)
  - logrotate (2.2.0)
  - cron (5.1.0)

<Ctrl-C>

[INFO] Cleanup
```

The cleanup part is handled using `trap` in the `DEPLOYMENT_CMD` script:

```
#!/usr/bin/env bash

set -o errexit
set -o errtrace
set -o pipefail
set -o nounset

function run {
  # <chef-client gets executed here>
}

function cleanup {
   # <do_stuff>
}

trap cleanup ERR SIGHUP SIGINT SIGTERM
```

If the same thing is executed via `knife ssh` then the signal is clearly not propagated:

```
$ knife ssh "QUERY" "DEPLOYMENT_CMD" --ssh-user XYZ --ssh-identity-file /path/to/.ssh/id_rsa --exit-on-error
[INFO] Starting chef-client wrapper
[INFO] <other things>
Starting Chef Client, version 14.11.21
resolving cookbooks for run list: [ ... ]
Synchronizing Cookbooks:
  - apt (7.0.0)
  - ulimit (1.0.0)
  - logrotate (2.2.0)
  - cron (5.1.0)

<Ctrl-C>
```

`trap` is never triggered in this scenario.

The only workaround that comes to my mind at this stage is a loop over `knife search "QUERY" -a name -F json` result and `ssh` execution for each host.

With `~/.ssh/config` support it's a matter of using this:

```
Host *
   RequestTTY force
```

## Hopping via multiple servers using different usernames

Let's say that I have to SSH via a gateway while executing `knife ssh`. That's what `--ssh-gateway` and `--ssh-gateway-identity` options are for. But what if I have to use `userA` when connecting to the gateway and `userB` when establishing outgoing connections later on?

Such issue can be resoled with a few `~/.ssh/config` entries.

---

#3602 is a bit ancient, but it's all about the same thing.

Here's a gist that could be helpful: https://gist.github.com/datagrok/dfe9604cb907523f4a2f

**Comment by scalp42** [comment]
@jwadolowski great insight 👌

**Comment by tas50** [comment]
We need to confirm if this has been fixed in the latest releases of knife bootstrap since that uses train and train will now use the ssh config.

**Comment by sanga1794** [comment]
@tas50, There are some fixes needed from chef side, i'm on it.

**Comment by tas50** [comment]
Closing this out since it's fixed in knife 17.4
